src/datasets/packaged_modules/text/text.py

```diff
@@ -53,43 +53,55 @@ def _split_generators(self, dl_manager):
     def _generate_tables(self, files):
         schema = pa.schema(self.config.features.type if self.config.features is not None else {"text": pa.string()})
         for file_idx, file in enumerate(files):
-            # open in text mode, by default translates universal newlines ("\n", "\r\n" and "\r") into "\n"
-            with open(file, encoding=self.config.encoding) as f:
+            # open in text mode:
+            # - By default (newline=None), it translates universal newlines ("\n", "\r\n" and "\r") into "\n"
+            # - If newline="", line endings are untranslated
+            with open(file, encoding=self.config.encoding, newline="" if self.config.newline else None) as f:
                 if self.config.sample_by == "line":
                     batch_idx = 0
+                    batch = ""
                     while True:
-                        batch = f.read(self.config.chunksize)
+                        new_batch = f.read(self.config.chunksize)
+                        batch += new_batch
                         if not batch:
                             break
-                        batch += f.readline()  # finish current line
-                        # StringIO.readlines, by default splits only on "\n" (and keeps line breaks)
-                        batch = StringIO(batch).readlines()
-                        if not self.config.keep_linebreaks:
-                            batch = [line.rstrip("\n") for line in batch]
-                        pa_table = pa.Table.from_arrays([pa.array(batch)], schema=schema)
+                        batch = batch.split(self.config.newline or "\n")
```

Comment:
If the batch ends on "\n", here you will end up with `'...\n'.split("\n")` which is equal to `['...', '']` which is maybe not expected